Assets/MixedRealityToolkit.Extensions/Socketer/Scripts/TCPConnectionManager.cs

```diff
@@ -0,0 +1,228 @@
+            client = SocketerClient.CreateSender(SocketerClient.Protocol.TCP, serverAddress, port);
+            client.Connected += OnClientConnected;
+            client.Disconnected += OnClientDisconnected;
+            client.Start();
+        }
+
+        private void OnServerConnected(SocketerClient client, int sourceId, string clientAddress)
+        {
+            Debug.Log("Server connected to " + clientAddress);
+            SocketEndpoint socketEndpoint = new SocketEndpoint(client, timeoutInterval, clientAddress, sourceId);
+            serverConnections[sourceId] = socketEndpoint;
+            socketEndpoint.QueueIncomingMessages(inputMessageQueue);
+            newConnections.Enqueue(socketEndpoint);
+        }
+
+        protected virtual void OnServerDisconnected(SocketerClient client, int sourceId, string clientAddress)
+        {
+            SocketEndpoint socketEndpoint;
+            if (serverConnections.TryRemove(sourceId, out socketEndpoint))
+            {
+                Debug.Log("Server disconnected from " + clientAddress);
+                socketEndpoint.StopIncomingMessageQueue();
+                oldConnections.Enqueue(socketEndpoint);
+            }
+        }
+
+        private void OnClientConnected(SocketerClient client, int sourceId, string hostAddress)
+        {
+            Debug.Log("Client connected to " + hostAddress);
+            SocketEndpoint socketEndpoint = new SocketEndpoint(client, timeoutInterval, hostAddress, sourceId);
+            clientConnection = socketEndpoint;
+            socketEndpoint.QueueIncomingMessages(inputMessageQueue);
+            newConnections.Enqueue(socketEndpoint);
+        }
+
+        private void OnClientDisconnected(SocketerClient client, int sourceId, string hostAddress)
+        {
+            if (clientConnection != null)
+            {
+                Debug.Log("Client disconnected");
+                clientConnection.StopIncomingMessageQueue();
+                oldConnections.Enqueue(clientConnection);
+                clientConnection = null;
+            }
+        }
+
+        private void Update()
+        {
+            DateTime utcNow = DateTime.UtcNow;
+            if (clientConnection != null)
+            {
+                clientConnection.CheckConnectionTimeout(utcNow);
+            }
+
+            foreach (SocketEndpoint endpoint in serverConnections.Values)
+            {
+                endpoint.CheckConnectionTimeout(utcNow);
+            }
+
+            SocketEndpoint connection;
+            while (newConnections.TryDequeue(out connection))
+            {
+                OnConnected?.Invoke(connection);
+            }
+
+            while (oldConnections.TryDequeue(out connection))
+            {
+                OnDisconnected?.Invoke(connection);
+            }
+
+            IncomingMessage resultPack = null;
+            while (inputMessageQueue.TryDequeue(out resultPack))
```

Comment:
C# now supports inline declarations for out variables.

Reply:
I think that message queueing shouldn't happen on the Unity thread.
A socket listener should happen on another thread, that listener should store incoming data in a threadsafe queue, which the unity thread polls on update. Otherwise incoming data can stall Unity's operations.

Reply:
Enqueues should be occurring off the main/UI thread. The Connected/Disconnected callbacks of the socket client aren't driven by an update call. Right now only dequeues should be conducted on the UI thread.